### ISSUE TITLE
ExceededMemoryLimitException at presto 0.148, 0.149

### ISSUE THREAD

**Issue by wyukawa** [prestodb/presto#5531]
Hi

I use presto 0.149 + jdk1.8.0_92 + HDP 2.1 + hive connector + CentOS 6.5.

If I execute the following query, ```Query exceeded max memory size of 70GB``` error occurs.
```
select * from test.hogeview where yyyymmdd='20160622' limit 1
```

test.hogeview is presto view and has many ```left outer join``` clause and one ```group by``` clause.

![](https://gyazo.com/d58bc564359ba4f248b85d755c425344.png)

Although same error occurs at presto 0.148, there is no error at presto 0.147.

worker config is the following.
```
coordinator=false
http-server.http.port=...
query.max-memory=70GB
query.max-memory-per-node=3GB
discovery.uri=...
task.cpu-timer-enabled=false
exchange.http-client.request-timeout=120s
```

**Comment by wyukawa** [comment]
at presto 0.147, query returns in 5 seconds and Total Memory is 46.51MB
![](https://gyazo.com/bcc7a381d0457013df5b156839f4c680.png)

but at presto 0.148, Total Memory is 68.07GB and ExceededMemoryLimitException occurs
![](https://gyazo.com/adf2b3a05d17d0a4fc8081a1afa56563.png)

**Comment by wyukawa** [comment]
Driver count is very different even if same query.

|presto ver|totaldrivers|completeddrivers|queueddrivers|runningdrivers|
|---|---|---|---|---|---|
|0.147|245|227|0|18|
|0.148|9747|7487|129|2131|

**Comment by dain** [comment]
Maybe predicate pushdown stopped working for your query. Can you post the explain plan from the two different systems?

-dain

**Comment by wyukawa** [comment]
Thank you for comment!

I check query exlain, I confirmed that predicate pushdown stopped working and caused full partition scan.

I can reproduce this problem even if relative simple query.

query
```
select * from test.test_view where yyyymmdd='20160701' limit 1
```

presto view ddl
```
create view test.test_view
as
select
	dt as yyyymmdd,
	format_datetime(date_parse(dt, '%Y%m%d'), 'yyyy-MM-dd') as yyyymmdd_hyphen,
    case when age_category is null then '0.Skip' else age_category end as age_category,
    count(id) as cnt
from
    (
    select
		dt,
        case when age between 16 and 18 then '1.16-18'
             when age between 19 and 22 then '2.19-22'
             when age between 23 and 29 then '3.23-29'
             when age between 30 and 39 then '4.30-39'
             when age between 40 and 49 then '5.40-49'
             when age between 50 and 59 then '6.50-59'
             when age >= 60 then '7.60-'
             else null
        end as age_category,
        id
    from
    (
        select
			a.dt,
            case when b.birthday is null then null
            else date_diff('year', date_parse(b.birthday, '%Y-%m-%d'), (date_parse(a.dt, '%Y%m%d') - INTERVAL '1' DAY)) end as age,
            a.id
        from
            test.hoge a
        LEFT OUTER JOIN
            test.piyo b
        ON
            a.id = b.piyoid
            and a.dt = b.dt
    ) tmp
) tmp2
group by 
	dt,
	age_category;
```

explain at presto 0.147
```
- Output[yyyymmdd, yyyymmdd_hyphen, age_category, cnt] => [dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint]
        yyyymmdd := dt
        yyyymmdd_hyphen := format_datetime
        age_category := expr_33
        cnt := count
    - Limit[1] => [dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint]
        - RemoteExchange[GATHER] => dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint
            - LimitPartial[1] => [dt:varchar, expr_33:varchar(7), count:bigint, format_datetime:varchar]
                - Project => [dt:varchar, expr_33:varchar(7), count:bigint, format_datetime:varchar]
                        expr_33 := (CASE WHEN ("expr_19" IS NULL) THEN CAST('0.Skip' AS VARCHAR(7)) ELSE "expr_19" END)
                        format_datetime := "format_datetime"("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)), CAST('yyyy-MM-dd' AS VARCHAR))
                    - Aggregate(FINAL)[dt, expr_19] => [dt:varchar, expr_19:varchar(7), $hashvalue:bigint, count:bigint]
                            count := "count"("count_71")
                        - RemoteExchange[REPARTITION] => dt:varchar, expr_19:varchar(7), count_71:bigint, $hashvalue:bigint
                            - Aggregate(PARTIAL)[dt, expr_19] => [dt:varchar, expr_19:varchar(7), $hashvalue_91:bigint, count_71:bigint]
                                    count_71 := "count"("id")
                                - Project => [dt:varchar, expr_19:varchar(7), id:bigint, $hashvalue_91:bigint]
                                        $hashvalue_91 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("dt"), 0)), COALESCE("$operator$hash_code"("expr_19"), 0))
                                    - Project => [dt:varchar, expr_19:varchar(7), id:bigint]
                                            expr_19 := (CASE WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '16' AND BIGINT '18') THEN '1.16-18' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '19' AND BIGINT '22') THEN '2.19-22' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '23' AND BIGINT '29') THEN '3.23-29' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '30' AND BIGINT '39') THEN '4.30-39' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '40' AND BIGINT '49') THEN '5.40-49' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '50' AND BIGINT '59') THEN '6.50-59' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) >= BIGINT '60') THEN CAST('7.60-' AS VARCHAR(7)) END)
                                        - LeftJoin[("id_72" = "hogeid_73") AND ("dt_74" = "dt_75")] => [id:bigint, dt:varchar, id_72:bigint, dt_74:varchar, $hashvalue_85:bigint, hogeid:bigint, dt_2:varchar, birthday:varchar, hogeid_73:bigint, dt_75:varchar, $hashvalue_88:bigint]
                                            - Project => [id:bigint, dt:varchar, id_72:bigint, dt_74:varchar, $hashvalue_85:bigint]
                                                - RemoteExchange[REPARTITION] => id:bigint, dt:varchar, id_72:bigint, dt_74:varchar, $hashvalue_85:bigint, $hashvalue_86:bigint
                                                    - Project => [id:bigint, dt:varchar, $hashvalue_87:bigint]
                                                            $hashvalue_87 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("id"), 0)), COALESCE("$operator$hash_code"("dt"), 0))
                                                        - TableScan[hive:hive:test:hoge, originalConstraint = ('20160701' = "dt")] => [id:bigint, dt:varchar]
                                                                LAYOUT: hive
                                                                id := HiveColumnHandle{clientId=hive, name=id, hiveType=bigint, hiveColumnIndex=0, partitionKey=false}
                                                                dt := HiveColumnHandle{clientId=hive, name=dt, hiveType=string, hiveColumnIndex=-1, partitionKey=true}
                                                                    :: [[20160701]]
                                            - Project => [hogeid:bigint, dt_2:varchar, birthday:varchar, hogeid_73:bigint, dt_75:varchar, $hashvalue_88:bigint]
                                                - RemoteExchange[REPARTITION] => hogeid:bigint, dt_2:varchar, birthday:varchar, hogeid_73:bigint, dt_75:varchar, $hashvalue_88:bigint, $hashvalue_89:bigint
                                                    - Project => [birthday:varchar, $hashvalue_90:bigint, dt_2:varchar, hogeid:bigint]
                                                            $hashvalue_90 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("hogeid"), 0)), COALESCE("$operator$hash_code"("dt_2"), 0))
                                                        - Filter[(CAST('20160701' AS VARCHAR) = "dt_2")] => [hogeid:bigint, birthday:varchar, dt_2:varchar]
                                                            - TableScan[hive:hive:test:piyo, originalConstraint = ('20160701' = "dt_2")] => [hogeid:bigint, birthday:varchar, dt_2:varchar]
                                                                    LAYOUT: hive
                                                                    hogeid := HiveColumnHandle{clientId=hive, name=hogeid, hiveType=bigint, hiveColumnIndex=0, partitionKey=false}
                                                                    birthday := HiveColumnHandle{clientId=hive, name=birthday, hiveType=string, hiveColumnIndex=4, partitionKey=false}
                                                                    dt_2 := HiveColumnHandle{clientId=hive, name=dt, hiveType=string, hiveColumnIndex=-1, partitionKey=true}
                                                                        :: [[20160701]]
```

explain at presto 0.148
```
- Output[yyyymmdd, yyyymmdd_hyphen, age_category, cnt] => [dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint]
        yyyymmdd := dt
        yyyymmdd_hyphen := format_datetime
        age_category := expr_33
        cnt := count
    - Limit[1] => [dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint]
        - RemoteExchange[GATHER] => dt:varchar, format_datetime:varchar, expr_33:varchar(7), count:bigint
            - LimitPartial[1] => [dt:varchar, expr_33:varchar(7), count:bigint, format_datetime:varchar]
                - Project => [dt:varchar, expr_33:varchar(7), count:bigint, format_datetime:varchar]
                        expr_33 := (CASE WHEN ("expr_19" IS NULL) THEN CAST('0.Skip' AS VARCHAR(7)) ELSE "expr_19" END)
                        format_datetime := "format_datetime"("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)), CAST('yyyy-MM-dd' AS VARCHAR))
                    - Aggregate(FINAL)[dt, expr_19] => [dt:varchar, expr_19:varchar(7), $hashvalue:bigint, count:bigint]
                            count := "count"("count_71")
                        - RemoteExchange[REPARTITION] => dt:varchar, expr_19:varchar(7), count_71:bigint, $hashvalue:bigint
                            - Aggregate(PARTIAL)[dt, expr_19] => [dt:varchar, expr_19:varchar(7), $hashvalue_76:bigint, count_71:bigint]
                                    count_71 := "count"("id")
                                - Project => [dt:varchar, expr_19:varchar(7), id:bigint, $hashvalue_76:bigint]
                                        $hashvalue_76 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("dt"), 0)), COALESCE("$operator$hash_code"("expr_19"), 0))
                                    - Project => [dt:varchar, expr_19:varchar(7), id:bigint]
                                            expr_19 := (CASE WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '16' AND BIGINT '18') THEN '1.16-18' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '19' AND BIGINT '22') THEN '2.19-22' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '23' AND BIGINT '29') THEN '3.23-29' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '30' AND BIGINT '39') THEN '4.30-39' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '40' AND BIGINT '49') THEN '5.40-49' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) BETWEEN BIGINT '50' AND BIGINT '59') THEN '6.50-59' WHEN ((CASE WHEN ("birthday" IS NULL) THEN CAST(null AS BIGINT) ELSE "date_diff"(CAST('year' AS VARCHAR), "date_parse"("birthday", CAST('%Y-%m-%d' AS VARCHAR)), ("date_parse"("dt", CAST('%Y%m%d' AS VARCHAR)) - "$literal$interval day to second"(BIGINT '86400000'))) END) >= BIGINT '60') THEN CAST('7.60-' AS VARCHAR(7)) END)
                                        - LeftJoin[("id" = "hogeid") AND ("dt" = "dt_2")] => [dt:varchar, id:bigint, $hashvalue_72:bigint, birthday:varchar, dt_2:varchar, hogeid:bigint, $hashvalue_74:bigint]
                                            - RemoteExchange[REPARTITION] => dt:varchar, id:bigint, $hashvalue_72:bigint
                                                - Project => [dt:varchar, $hashvalue_73:bigint, id:bigint]
                                                        $hashvalue_73 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("id"), 0)), COALESCE("$operator$hash_code"("dt"), 0))
                                                    - TableScan[hive:hive:test:hoge, originalConstraint = ('20160701' = "dt")] => [id:bigint, dt:varchar]
                                                            LAYOUT: hive
                                                            id := HiveColumnHandle{clientId=hive, name=id, hiveType=bigint, hiveColumnIndex=0, partitionKey=false}
                                                            dt := HiveColumnHandle{clientId=hive, name=dt, hiveType=string, hiveColumnIndex=-1, partitionKey=true}
                                                                :: [[20160701]]
                                            - RemoteExchange[REPARTITION] => birthday:varchar, dt_2:varchar, hogeid:bigint, $hashvalue_74:bigint
                                                - Project => [birthday:varchar, $hashvalue_75:bigint, dt_2:varchar, hogeid:bigint]
                                                        $hashvalue_75 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("hogeid"), 0)), COALESCE("$operator$hash_code"("dt_2"), 0))
                                                    - TableScan[hive:hive:test:piyo, originalConstraint = true] => [hogeid:bigint, birthday:varchar, dt_2:varchar]
                                                            LAYOUT: hive
                                                            hogeid := HiveColumnHandle{clientId=hive, name=hogeid, hiveType=bigint, hiveColumnIndex=0, partitionKey=false}
                                                            birthday := HiveColumnHandle{clientId=hive, name=birthday, hiveType=string, hiveColumnIndex=4, partitionKey=false}
                                                            dt_2 := HiveColumnHandle{clientId=hive, name=dt, hiveType=string, hiveColumnIndex=-1, partitionKey=true}
                                                                :: [[20150308, 20160706]]
```

**Comment by wyukawa** [comment]
Hi

I can simplify query more.

for example, aaa and bbb table have partition column dt.

the following query scan full bbb table at presto 0.148. but at presto 0.147, full scan doesn't occur.
```
SELECT
	*
FROM
    aaa a
LEFT OUTER JOIN
    bbb b
ON
    a.id = b.id
    and a.dt = b.dt
WHERE
  a.dt='20160701'
LIMIT 1
```

**Comment by martint** [comment]
@geraint0923, can you investigate this?

**Comment by martint** [comment]
@losipiuk, this may be related to the outer join changes

**Comment by cberner** [comment]
Thanks for narrowing this down! I suspect this is related to this item in the 0.148 release: `Improve outer join predicate push down.` CC: @losipiuk

**Comment by cawallin** [comment]
@losipiuk is out for the next few weeks, but I'll see if anyone else has enough context on the code to take a look

**Comment by geraint0923** [comment]
@martint I am looking into it.

**Comment by kokosing** [comment]
@geraint0923 I have some (not much though ;/) context about this in case you need some help

**Comment by geraint0923** [comment]
After investigation, it seems that `processLimitedOuterJoin ` (https://github.com/prestodb/presto/blob/ddfeb541dcf8be8cad3ec64be37c7406d2de366b/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java#L409) doesn't add `b.dt='20160701'` which is a equality combining join predicate and inherited predicate into `innerPushdownConjuncts`.

I am working on a fix now.

**Comment by losipiuk** [comment]
@geraint0923, thanks for looking into it. I have somewhat limited internet access here, but I can take a look at this now if you do not have a fix yet.

**Comment by losipiuk** [comment]
Just in case you did not have time to work on this one.
Here is the PR which fixes issue and hopefully does not brake anything:
https://github.com/Teradata/presto/pull/250

This code actually was originally in `PredicatePushDown.java` before refactor. And seems to be there to exactly handle cases like the one from this issue.

I do not see a reason why it would break something - but this is possible that I am missing something. 
I tested the code with clone of the original query - rewritten to tpch schema:

```sql
SELECT * FROM nation LEFT OUTER JOIN region ON nation.regionkey = region.regionkey and nation.name = region.name WHERE nation.name = 'dupa' LIMIT 1;
```

And plan looks right. Here is the join inner side part:

```
- Project => [name_1:varchar(25), regionkey_0:bigint, comment_2:varchar(152), regionkey_40:bigint, name_42:varchar(25), $hashvalue_57:bigint]
                                 - RemoteExchange[REPARTITION] => name_1:varchar(25), regionkey_0:bigint, comment_2:varchar(152), regionkey_40:bigint, name_42:varchar(25), $hashvalu
                                     - Project => [comment_2:varchar(152), $hashvalue_59:bigint, name_1:varchar(25), regionkey_0:bigint, $hashvalue_60:bigint]
                                             $hashvalue_59 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("regionkey_0"), 0)), COALESCE("$operator$hash_
                                             $hashvalue_60 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("name_1"), 0)), COALESCE("$operator$hash_code"
                                         - Filter[(("name_1" = CAST('dupa' AS VARCHAR(25))) AND (CAST('dupa' AS VARCHAR(25)) = "name_1"))] => [regionkey_0:bigint, name_1:varchar(25)
                                             - TableScan[tpch:tpch:region:sf0.01, originalConstraint = ('dupa' = "name_1")] => [regionkey_0:bigint, name_1:varchar(25), comment_2:var
                                                     regionkey_0 := tpch:regionkey
                                                     name_1 := tpch:name
                                                     comment_2 := tpch:comment
```

The only issue I see here is that we have the conjunct `(("name_1" = CAST('dupa' AS VARCHAR(25)))` repeated twice with reversed order of arguments. This does not break anything but is not best for performance. Though it does not seem related to this issue.

**Comment by kokosing** [comment]
I added a test to the fix proposed by @losipiuk, so actual PR with a fix is here: #5618

**Comment by geraint0923** [comment]
@losipiuk @kokosing Thank you for your help! The PR seems reasonable and I tested it with my test case and it now has the correct pushdown.
